### PR TITLE
Change modis_aod2ioda.py to v2

### PR DESCRIPTION
**Pull request by andytangborn**
## Description

*(Instructions: replace text in this and all sections with your own text)*

Provide a detailed description of what this PR does.

What problem does it fix? What new capability does it add?

[See the JEDI Documentation](https://jointcenterforsatellitedataassimilation-jedi-docs.readthedocs-hosted.com/en/latest/inside/practices/pullrequest.html) for further information.

### Issue(s) addressed

Link the issues to be closed with this PR
- fixes #<issue_number>

If this does not close an existing issue, then [be sure to add a Pipeline, Label, Estimate, Assignees, and Epic](https://jointcenterforsatellitedataassimilation-jedi-docs.readthedocs-hosted.com/en/latest/inside/practices/issues.html)

## Acceptance Criteria (Definition of Done)

What does it mean for this PR to be finished?

## Dependencies

If there are PRs that need to be merged before or along with this one, please add "waiting for another PR" label and list the dependencies in the other repositories (example below). Note that the branches in the other repositories should have matching names for the automated tests to pass.
Waiting on the following PRs:
- [ ] waiting on JCSDA/eckit/pull/<pr_number>
- [ ] waiting on JCSDA/atlas/pull/<pr_number>

## Impact

Please list the other repositories (if any) that this PR will require changes in (example below)

- [ ] saber
- [ ] ufo
- [ ] ...

### CODE REVIEW

**Review by rhoneyager — DISMISSED** [review]
Minor changes are noted above.

Also, ```ObsError/aerosol_optical_depth``` has a different data type than ```ObsValue/aerosol_optical_depth``` (double vs float). Please change the ObsError values so that they are 32-bit floats.

**Comment by rhoneyager** [comment]
@andytangborn I am updating this PR to bring it in sync with develop. Can you regenerate the test reference file?

**Comment by andytangborn** [comment]
sure. I assume this is because of the change in size of the test reference.
So I'll need to pull this new version.

**Comment by andytangborn** [comment]
I pushed a new version with the updated reference.

File: src/chem/modis_aod2ioda.py [hunk 1]
```diff
@@ -37,27 +37,38 @@
...
-            self.varDict[iodavar]['errKey'] = iodavar, self.writer.OerrName()
-            self.varDict[iodavar]['qcKey'] = iodavar, self.writer.OqcName()
+            self.varDict[iodavar]['valKey'] = iodavar, iconv.OvalName()
+            self.varDict[iodavar]['errKey'] = iodavar, iconv.OerrName()
+            self.varDict[iodavar]['qcKey'] = iodavar, iconv.OqcName()
+            self.varAttrs[iodavar, iconv.OvalName()]['coordinates'] = 'longitude latitude'
+            self.varAttrs[iodavar, iconv.OerrName()]['coordinates'] = 'longitude latitude'
+            self.varAttrs[iodavar, iconv.OqcName()]['coordinates'] = 'longitude latitude'
+            self.varAttrs[iodavar, iconv.OvalName()]['units'] = 'unitless'
+            self.varAttrs[iodavar, iconv.OerrName()]['units'] = 'unitless'
```
Comment (srherbener) [inline src/chem/modis_aod2ioda.py:71]: After some discussion, we settled on "unitless" being the units name for categorical variables such as qc flags and enumerations. Whereas for physical quantities that happen to be without units, we settled on using '1' as the units name. We are doing this so we can distinguish between variables that can be used in mathematical calculations (units = "1") from those that cannot (units = "unitless").

I'm thinking that aerosol_optical_depth is a unit-less physical quantity that can be used in mathematical formulas. If that is correct, then please set the units to '1'.

Also, please add in the OqcName version of the variables and set their units to "unitless". Eg,
```
self.varAttrs[iodavar, iconv.OqcName()]['units'] = 'unitless'
```

Thanks!

File: src/chem/modis_aod2ioda.py [hunk 2]
```diff
@@ -37,27 +37,38 @@
...
         for ncvar, iodavar in obsvars.items():
-            self.varDict[iodavar]['valKey'] = iodavar, self.writer.OvalName()
-            self.varDict[iodavar]['errKey'] = iodavar, self.writer.OerrName()
-            self.varDict[iodavar]['qcKey'] = iodavar, self.writer.OqcName()
+            self.varDict[iodavar]['valKey'] = iodavar, iconv.OvalName()
+            self.varDict[iodavar]['errKey'] = iodavar, iconv.OerrName()
+            self.varDict[iodavar]['qcKey'] = iodavar, iconv.OqcName()
+            self.varAttrs[iodavar, iconv.OvalName()]['coordinates'] = 'longitude latitude'
+            self.varAttrs[iodavar, iconv.OerrName()]['coordinates'] = 'longitude latitude'
+            self.varAttrs[iodavar, iconv.OqcName()]['coordinates'] = 'longitude latitude'
```
Comment (srherbener) [inline src/chem/modis_aod2ioda.py:69]: If the data provider has defined missing values for any of these variables, please specify those using the _FillValue variable attribute. Use "_FillValue" for the attribute name, and use the same data type as the variable for the attribute data type. Thanks!

**Review by srherbener — CHANGES_REQUESTED** [review]
In general, the code looks great. Just a couple comments about the variable attributes. Thanks!

**Comment by andytangborn** [comment]
I've merged develop into this branch and pulled again. I also made the changes above for units and FillValue.  I'm able to make, but a number of ctests are failing now. It looks mainly like the testoutput files don't match the testrun files for a number of the ioda converters.  And it seems to be primarily nlocs=0 vs 1,2,3 ...     Am I missing something here?

**Comment by srherbener** [comment]
> I've merged develop into this branch and pulled again. I also made the changes above for units and FillValue. I'm able to make, but a number of ctests are failing now. It looks mainly like the testoutput files don't match the testrun files for a number of the ioda converters. And it seems to be primarily nlocs=0 vs 1,2,3 ... Am I missing something here?

A recent PR to the ioda repository changed the default nlocs variable initialization from 1..n to all zeros. This was done to help compression (since we aren't using the nlocs values yet). So the all zeros is expected and you should be able to simply copy your test output in `build_directory/iodaconv/test/testrun/output_file` to the test reference `source_directory/iodaconv/test/testoutput/output_file`.

**Comment by andytangborn** [comment]
I've fixed most of the ctest errors by copying the files in testrun to testoutput.  I'm not sure why this was needed, because I would have expected them to be updated when I merged developed into my branch, and did 'git pull'.  Also, there remains one ctest failing that I don't understand. I'm getting this error: 
Command: "/scratch1/NCEPDEV/da/Andrew.Tangborn/JEDI/codesprint_final/build/bin/iodaconv_comp.sh" "netcdf" "/scratch2/NCEPDEV/marineda/Jong.Kim/anaconda3-save/bin/python3.7 /scratch1/NCEPDEV/da/Andrew.Tangborn/JEDI/codesprint_final/build/bin/ncep_classes.py
                          -p testinput/
                          -i b001xx007.20200310.bufr
                          -o testrun/ioda.NC001007.2020031012.nc
                          -d 2020031012
                          -m 5
                          -l NC001007.yaml
                          -ot NC001007" "ioda.NC001007.2020031012.nc" "0.0"
Directory: /scratch1/NCEPDEV/da/Andrew.Tangborn/JEDI/codesprint_final/build/iodaconv/test
"test_iodaconv_ncep_bufr" start time: Nov 17 18:43 UTC
Output:
----------------------------------------------------------
RuntimeError: module compiled against API version 0xe but this version of numpy is 0xd
Traceback (most recent call last):
  File "/scratch1/NCEPDEV/da/Andrew.Tangborn/JEDI/codesprint_final/build/bin/ncep_classes.py", line 3, in <module>
    import ncepbufr
  File "/scratch1/NCEPDEV/jcsda/jedipara/opt/modules/intel-2020.2/bufr/noaa-emc-11.5.0/lib/python3.7/site-packages/ncepbufr/__init__.py", line 1, in <module>
    import _bufrlib
ImportError: numpy.core.multiarray failed to import
<end of output>
Test time =   2.04 sec
----------------------------------------------------------
Test Failed.

**Comment by andytangborn** [comment]
I've updated the reference file, and the ctest is working for me. But I'm getting an error message when I try to run the converter outside of the ctest.   I think the error is related to the python version that I am loading: 

  File "/scratch1/NCEPDEV/da/Andrew.Tangborn/JEDI/codesprint_final/build/lib/pyiodaconv/ioda_conv_engines.py", line 2, in <module>
    import ioda_obs_space as ioda_os
ModuleNotFoundError: No module named 'ioda_obs_space'  

I'm using:

module use -a /scratch2/NCEPDEV/marineda/Jong.Kim/save/modulefiles/
module load anaconda/3.15.1

**Comment by CoryMartin-NOAA** [comment]
> ModuleNotFoundError: No module named 'ioda_obs_space'

This is because you need to add IODA to your `$PYTHONPATH`.

It should be something like:
`export PYTHONPATH=$PYTHONPATH:/path/to/bundle/build/lib/python3.X/pyioda/`

**Comment by andytangborn** [comment]
Ok. I'll try that.

**Comment by srherbener** [comment]
@andytangborn the test failures are due to a recent PR to ioda where the default initialization for dimension variables was changed from 1..n to all zeros. This is the only difference between the test output (`build_area/iodaconv/test/testrun/modis_aod.nc`) and the test reference (`build_area/iodaconv/test/testoutput/modis_aod.nc`), so it's okay to copy the test output to the test reference. Thanks!

File: src/chem/modis_aod2ioda.py (hunk 1 above)
Comment (srherbener) [reply src/chem/modis_aod2ioda.py:74]: Still need to add the units to the "OqcName" version of the variable as shown above. Thanks!

**Comment by andytangborn** [comment]
Ok. I'll do that, and also fix the default initialization fordimension
variables as well.

File: src/chem/modis_aod2ioda.py (hunk 1 above)
Comment (srherbener) [reply src/chem/modis_aod2ioda.py:75]: Thanks for adding this in!

**Review by srherbener — APPROVED** [review]

**Comment by srherbener** [comment]
CI test failures are down to two tests with easy fixes:
- iodaconv_chem_coding_norms
    - Extra white space at end of a line in the modis aod script
- test_iodaconv_modis_aod
    - Need to update the test reference file to accommodate the recent ioda PR that switched the default initialization of dimensions from 1..n to all zeros

Thanks for taking care of this!

**Comment by andytangborn** [comment]
Stephen,

The first one is easy, but for the second item, I'm not sure if I need to
simply copy the test reference file to testoutput, or do I need to make a
code change. Currently the version I have still outputs 1,...,n  for nlocs.

**Comment by CoryMartin-NOAA** [comment]
@andytangborn does that test fail for you locally or does it pass? Perhaps you need to link to the newest version of IODA before running your test and then copy the test output file.

**Comment by andytangborn** [comment]
Do you mean merge develop into my branch, and then pull?

**Comment by CoryMartin-NOAA** [comment]
> Do you mean merge develop into my branch, and then pull?
> […](#)

I mean for `ioda` not `ioda-converters` since that is where the nlocs change occurred. I suggest: 1) making sure `ioda` is set to `develop` in your CMakeLists.txt, 2) run `make update` in your build directory, 3) run `make -j8` in your build directory and then 4) rerun your `ioda-converters` test.

**Comment by andytangborn** [comment]
Doing this gives an error message

*" /usr/bin/git checkout -q develop*

*    error: Your local changes to the following files would be overwritten
by checkout:"*

**Comment by CoryMartin-NOAA** [comment]
@andytangborn on which repository? If you want, feel free to schedule a 1 on 1 for this afternoon to sort this out.

**Comment by andytangborn** [comment]
This is at the ioda-bundle level.  Is that where you meant CMakeLists.txt
to be pointed to "develop"?

**Comment by andytangborn** [comment]
Getting an error message during build:

/scratch1/NCEPDEV/da/Andrew.Tangborn/JEDI/codesprint_final/src/ioda-bundle/oops/src/oops/base/ObsSpaces.h(98):
error: identifier "validateAndDeserialize" is undefined


validateAndDeserialize<Parameters_>(util::vectoriseAndFilter(conf,
"obs space")),

                ^

          detected during:

            instantiation of "int ioda::TimeIodaIO<MODEL>::execute(const
eckit::Configuration &) const [with MODEL=ioda::IodaTrait]" at line 43 of
"/scratch1/NCEPDEV/da/Andrew.Tangborn/JEDI/codesprint_final/build/ioda/include/ioda/mains/timeIodaIO.h"

            instantiation of "ioda::TimeIodaIO<MODEL>::TimeIodaIO(const
eckit::mpi::Comm &) [with MODEL=ioda::IodaTrait]" at line 19 of
"/scratch1/NCEPDEV/da/Andrew.Tangborn/JEDI/codesprint_final/src/ioda-bundle/ioda/src/mains/timeIodaIO.cc"

**Comment by CoryMartin-NOAA** [comment]
@andytangborn Do a `make update` in the build directory and try again.
This was a bug introduced today that has been fixed as of 5 mins ago: https://github.com/JCSDA-internal/oops/pull/1567

**Comment by andytangborn** [comment]
I'm getting a lot of ctest failures due to differences in the output files.
I would have thought that git update would have downloaded the latest
testoutput files.  But I think that all I should worry about is modis_aod.nc,
since that's the one I've changed.

**Comment by CoryMartin-NOAA** [comment]
@andytangborn did you do the following on your branch:
```
git checkout develop
git pull
git checkout feature/modis_aod_upgrade
git merge develop
```
?

**Comment by andytangborn** [comment]
I didn't merge.

**Review by jeromebarre — APPROVED** [review]

**Comment by YoulongXia-NOAA** [comment]
@CoryMartin-NOAA @rhoneyager @srherbener, thank you giving me such a chance and I can make a practice.  Your guidance and instructions are very appreciated.  Have a good weekend. Hopefully you can merge it.

**Comment by srherbener** [comment]
@rhoneyager you mentioned a need to re-review. Since this PR has been open for a long time, will it be possible to do that review soon? Thanks!

**Comment by andytangborn** [comment]
This would be good to do. I can work on changes if needed.  I'm also
working on viirs to add the date stamp.

**Pull request merged**